Model/ApiPayloadAuth.php

```diff
@@ -0,0 +1,343 @@
+use Mautic\CoreBundle\Model\AbstractCommonModel;
+use MauticPlugin\MauticContactClientBundle\Entity\Auth;
+use MauticPlugin\MauticContactClientBundle\Entity\AuthRepository;
+use MauticPlugin\MauticContactClientBundle\Entity\ContactClient;
+use MauticPlugin\MauticContactClientBundle\Helper\TokenHelper;
+
+/**
+ * Class ApiPayloadAuth.
+ */
+class ApiPayloadAuth extends AbstractCommonModel
+{
+    /** @var TokenHelper */
+    protected $tokenHelper;
+
+    /** @var bool */
+    protected $test;
+
+    /** @var array */
+    protected $requiredPayloadTokens;
+
+    /** @var array */
+    protected $operations;
+
+    /** @var ContactClient */
+    protected $contactClient;
+
+    /** @var array */
+    protected $authOperations = [];
+
+    /** @var EntityManager */
+    protected $em;
+
+    /** @var array */
+    protected $authRequestOperations;
+
+    /** @var AuthRepository */
+    protected $authRepository;
+
+    /** @var array */
+    protected $previousPayloadAuthTokens = [];
+
+    /**
+     * ApiPayloadAuth constructor.
+     *
+     * @param TokenHelper   $tokenHelper
+     * @param EntityManager $em
+     */
+    public function __construct(TokenHelper $tokenHelper, EntityManager $em)
+    {
+        $this->tokenHelper = $tokenHelper;
+        $this->em          = $em;
+
+        return $this;
+    }
+
+    /**
+     * @param $test
+     *
+     * @return $this
+     */
+    public function setTest($test)
+    {
+        $this->test = $test;
+
+        return $this;
+    }
+
+    /**
+     * @param $operations
+     *
+     * @return $this
+     */
+    public function setOperations($operations)
+    {
+        $this->operations = $operations;
+
+        return $this;
+    }
+
+    /**
+     * @param $contactClient
+     *
+     * @return $this
+     */
+    public function setContactClient(ContactClient $contactClient)
+    {
+        $this->contactClient = $contactClient;
+
+        return $this;
+    }
+
+    /**
+     * Reset local class variables.
+     *
+     * @param array $exclusions optional array of local variables to keep current values
+     *
+     * @return $this
+     */
+    public function reset($exclusions = ['tokenHelper', 'em'])
+    {
+        foreach (array_diff_key(
+                     get_class_vars(get_class($this)),
+                     array_flip($exclusions)
+                 ) as $name => $default) {
+            $this->$name = $default;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Step forward through operations to find the one we MUST start with.
+     *
+     * @return array
+     */
+    public function getStartOperation()
+    {
+        $id = 0;
+        if ($this->hasAuthRequest()) {
+            // Step through operations, skipping any at the beginning that can be skipped.
+            foreach ($this->operations as $id => $operation) {
+                if ($id == count($this->operations) - 1) {
+                    // This is the last request so it cannot be skipped.
```

Comment:
I dont know if its a valid scenario, but what if the last op is also the first op (ie, only one op defined) and its an Auth with stored value? I dont think this would actually validate in setup, but just asking.

Reply:
If the last op is the first op, the user never get's to this point. Nothing is skippable, no special treatment for auth.